src/Storages/MergeTree/MergeTreeWhereOptimizer.cpp

```diff
@@ -37,6 +37,7 @@ MergeTreeWhereOptimizer::MergeTreeWhereOptimizer(
     : table_columns{ext::map<std::unordered_set>(
         metadata_snapshot->getColumns().getAllPhysical(), [](const NameAndTypePair & col) { return col.name; })}
     , queried_columns{queried_columns_}
+    , primary_key_columns{metadata_snapshot->getPrimaryKey().column_names}
```

Comment:
You need the sorting key here. Sorting key is used as key in special `MergeTree` engines and may differ from the primary key.